image-registry/automate-pruning.md

```diff
@@ -0,0 +1,252 @@
+So that etcd storage is not overwhelmed by unused imagestreams
+And the cluster remains healthy
+
+#### Prune images on the registry
+
+As an OpenShift cluster admin
+I want the cluster to automatically prune unused images stored in the registry
+So that registry storage is managed
+And I don’t pay for storage consumed by images my organization doesn’t use
+
+#### Orphaned blobs
+
+As an OpenShift cluster admin
+I want the cluster to automatically remove orphaned blobs stored in the registry
+So that registry storage is managed
+And I don’t pay for storage consumed by images my organization doesn’t use
+
+### Implementation Details/Notes/Constraints
+
+The implementation relies the current image pruning process invoked by the
+`oc adm prune` command. Running this in a container can be accomplished by
+using the `cli` imagestream installed by the samples operator. The image must
+be run using a service account with the `system:image-pruner` cluster role.
+
+The image registry operator will be updated to install the following:
+
+1. A `ServiceAccount` dedicated to the image pruner.
+2. A `ClusterRoleBinding` granting the `system:image-pruner` role to the pruner
+service account.
+3. A `CronJob` which runs the image pruner with the following tunable parameters:
+   1. Schedule for `CronJob`
+   2. `--keep-tag-revision` command line argument for the pruner
+   3. `--keep-younger-than` command line argument for the pruner
+   4. Resources for the `CronJob` pod template
+
+### Risks and Mitigations
+
+#### Resource consumption
+
+Risk: Image pruning requires siginficant memory to compute the blob layer graph
+needed to identify "prunable" layers.
+
+Mitigation: API will allow cluster admins to adjust the resources allocated to
+the `CronJob`.
+
+Risk: Image pruning requires too much memory to be scheduled to a node
+Risk: Image pruning exhausts resources used for application workloads
+
+Mitigation: API will allow admins to use standard tools to let pruning be run
+on desired machine pools (`affinity`, `tolerations`, `nodeSelector`)
+
+##### Concurrent pruning jobs
+
+Risk: Pruning jobs cannot run concurrently.
+
+Mitigation: The `CronJob` for pruning must be configured with
+`concurrencyPolicy: Forbid`
+
+#### Removal of past pruning jobs
+
+Risk: Runs of the pruning job create another set of objects in etcd that need
+to be pruned.
+
+Mitigation: `CronJob` specifies a history limit for successful and failed jobs.
+This will be a tunable parameter via the API.
+
+#### Pruning of in-use images
+
+Risk: Pruner is not aware of a particular image, resulting in it being pruned
+prior to use.
+
+Mitigation: None at present - this is an inherent problem within the current
+pruning process.
+
+## Design Details
+
+### API
+
+Proposed changes to the image registry operator's config spec:
+
+```yaml
+spec:
+  ...
+  pruning:
+    schedule: "*/1 * * * *"
+    keepTagRevisions: 3
+    keepYoungerThan: 60m
+    resources: {}
```

Comment:
Something to consider - setting the resources for the pruner based on cluster info (number of `Image` objects, for example).

I don't think this is a requirement for GA, but would be a good feature to add to the backlog.

Reply:
Added to JIRA: https://jira.coreos.com/browse/DEVEXP-441